[#13] 회원 관련 코드 보완

canlion opened the pull request:
## 목표
 
* 회원 관련 코드 보완
  * 회원 속성 검증 로직 추가
  * 회원 정보 수정 기능 추가

### 회원 속성 검증 로직 추가
* bean validation 추가
  ```java
  public class MemberJoinDto {
    @NotBlank
    @Size(min = 8, max = 12)
    private String loginId;
    @NotBlank
    @Size(min = 8, max = 12)
    private String password;
    @NotBlank
    @Email
    private String email;
  }
   ```
* controller advice를 통해 `MethodArgumentNotValidException` 예외 발생 처리, 검증 결과를 응답으로 반환
   ```json
    {
        "errors": [
            {
                "objectError": null,
                "fieldName": "loginId",
                "message": "아이디는 8자 이상, 12자 이하여야 합니다."
            },
            {
                "objectError": null,
                "fieldName": "password",
                "message": "비밀번호는 8자 이상, 12자 이하여야 합니다."
            },
            {
                "objectError": null,
                "fieldName": "email",
                "message": "이메일 형식을 확인해주세요."
            }
        ]
    }
   ```

### 회원 정보 수정 기능 추가

* email, password 수정 기능 추가
* 두 필드가 모두 null인 경우 예외로 취급, 검증 결과를 응답으로 반환

---

# 240907 설정 초기화 및 코드 재작성
- 코드 작성이 막히는 것은 프로젝트에 대한 그림이 명확하지 않기 때문이라는 조언을 해주셔서 제가 생각하는 서비스를 사용자 시점에서 생각해보며 다시 구상을 하였고 필요한 기능들을 정리하였습니다.
- DB, validation 부분에서 너무 코드 복붙에 의존하였고, 회원 관련 부분만 쭉 진행하되 충분한 양을 작성하지 못해 먼저 제가 할 수 있는 부분부터 충분히 코드 작성을 하고 개선하고자 했습니다.
-  github copilot을 사용했습니다.

---

# 서비스 구상

## 회원 관련 API
### 엔티티
* Long memberId
* MemberType memberType
* String loginId
* String password
* String email
* String nickname
### MemberType - enum
* CUSTOMER - 소비자 - 리뷰 작성 가능
* OWNER - 판매자 - 가게 관리 가능
### API
#### 회원 가입
* `/members`
	* `POST`
		* input - `MemberJoinRequestDto`
			* MemberType memberType
			* String loginId
			* String password
			* String email
			* String nickname
		* output - ~~`MemberJoinResponseDto`~~ / `MemberInfoDto`
			* MemberType memberType
			* String loginId
			* String email
			* String nickname
* 중복 확인
	* `GET`
		* `/members/check-id/{loginId}`
		* `/members/check-nickname/{nickname}`
		* `/members/check-email/{email}`
	* output
		* boolean available
			* ex) `{"available": true}`
#### 로그인 & 로그아웃
* `/members/login`
	* `POST` - 회원식별번호(MemberId)를 세션에 저장
		* input - `MemberLoginDto`
			* String loginId
			* String password
		* output - cookie
			* boolean login
				* ex) `{"login": true}`
* `/members/logout`
	* `GET`
		* 세션 제거
#### 회원 정보 수정, 삭제
세션으로 인증, 로그인되지 않은 경우 `401 Unauthorized` 반환
 * 기존 정보 조회
	 * `/members/my-info`
		 * `GET`
			 * output - `MemberInfoDto`
				 * MemberType memberType
				 * String loginId
				 * String email
				 * String nickname
 * 정보 수정
	 * `PATCH`
		 * `/members/my-info/password`
		 * `/members/my-info/email`
		 * `/members/my-info/nickname`
	 * input - `MemberChangeableInfoDto` - 바꾸려는 정보 외에는 null
		* String password;  
		* String email;  
		* String nickname;
	 * output - status code OK
 * 회원 삭제
	 * `/members/my-info`
		 * `DELETE`
			 * output - `MemberInfoDto`

## 가게 관련
### 엔티티
* Long storeId
* String storeName
* String address
* String call
* String introduction

### API
#### 가게 등록
* `/stores`
	* `POST` - 어떤 멤버가 등록하는지는 세션의 memberId를 찾아서 활용
		* input - `StoreInfoDto`
			* String storeName
			* String address
			* String call
			* String introduction
		* output - `StoreInfoDto`
			* String storeName
			* String address
			* String call
			* String introduction
#### 가게 조회
* `/stores`
	* `GET`: 필터링 조건을 만족하는 가게 조회
		* 위치에 의한 필터링 기능, 음식에 의한 필터링 기능 필요
			* 음식 관련 기능 작성이 선행되어야 함
* `/stores/{storeId}`
	* `GET`: 개별 가게 정보 조회
		* output - `StoreInfoDto`
* `/stores/my-stores`
	* `GET`: 내가 소유한 가게 조회 - OWNER 타입의 회원만 사용 가능, 세션을 통한 인증 필요
		* output - `List<StoreInfoDto>`
#### 가게 정보 수정, 삭제
세션을 통한 인증 필요
* `/stores/{storeId}`
	* `PUT`: 회원 가입과는 다르게 통째로 업데이트
		* input - `StoreInfoDto`
		* output - `StoreInfoDto
	* `DELETE`: 가게 삭제
		* output - `StoreInfoDto`

canlion committed:
feat: 회원 추가 과정에 필드 검증 수행

- MemberJoinDto에 bean validation 적용
  - 검증 대상
    - loginId, password: NotBlank, Size(min = 8, max = 12)
    - email: NotBlank, Email
  - 검증에 실패한 경우 BadRequest와 함께 검증 실패 내역 반환
- bean validation 테스트 위해 MemberController 단위 테스트 코드 추가

canlion committed:
feat: 회원 정보 수정 기능 추가

- 회원 정보 수정 기능 추가
  - email, password 수정
  - 두 필드가 모두 null인 경우 BadRequest 처리

canlion committed:
test: 회원 정보 수정 테스트 추가

f-lab-lyan commented on api/src/main/java/burgermap/controller/MemberController.java:
```diff
@@ -55,4 +68,37 @@ public MemberResponseDto deleteMember(@PathVariable("memberId") Long memberId, H
+            @RequestBody @Valid MemberUpdateDto memberUpdateDto,
+            BindingResult bindingResult
+            ) throws MethodArgumentNotValidException, NoSuchMethodException {
+        if (memberUpdateDto.getEmail() == null && memberUpdateDto.getPassword() == null) {
```
두 개 이상의 필드에 대해서 비교하는건 비지니스 로직에 더 가깝다고 판단하는 것이 좋을 것 같네요.

f-lab-lyan commented on api/src/main/java/burgermap/controller/MemberController.java:
```diff
@@ -55,4 +68,37 @@ public MemberResponseDto deleteMember(@PathVariable("memberId") Long memberId, H
+            );
+        }
+
+        MemberResponseDto memberResponseDto = memberService.updateMember(memberId, memberUpdateDto);
```
Service가 DTO를 알아야 하나요?

f-lab-lyan commented on api/src/main/java/burgermap/repository/MySqlMemberRepository.java:
```diff
@@ -42,6 +43,21 @@ public Member deleteMember(Long memberId) {
+            if (memberUpdateDto.getEmail() != null) {
+                member.setEmail(memberUpdateDto.getEmail());
+            }
+            em.flush();
```
왜, em.flush()를 부르면, member가 저장되는건가요?

canlion committed:
chore: 회원 관련 코드 재작성을 위해 코드 및 설정 초기화

canlion committed:
refactor: 회원 관리 기능에 로그인 기능을 추가하여 재작성

* 기존 코드 재작성
  - 사유: 서비스의 기존 구상이 허술하여 기능 부족 및 코드 일관성 부족
  - 개선점:
      - 회원 타입 추가: CUSTOMER(소비자), OWNER(햄버거가게소유자)
      - 세션을 이용한 로그인 기능 추가
      - 회원가입을 위한 이메일, 닉네임, 아이디 중복 확인 기능 추가

canlion committed:
refactor: 역할이 중복되는 Dto 제거

* MemberResponseDto 제거
  - MemberInfoDto와 동일한 역할
    - 더 범용적인 명칭의 MemberInfoDto를 유지

canlion committed:
feat: 가게 관리 기능 추가

* 가게(Store) 관리 기능 추가
  - OWNER 회원에 의한 가게 추가
  - 특정 가게 조회
  - OWNER 회원이 소유한 모든 가게 조회
  - OWNER 회원이 소유한 가게 정보 수정
  - OWNER 회원이 소유한 가게 삭제

f-lab-lyan commented on api/src/main/java/burgermap/controller/MemberController.java:
```diff
@@ -28,31 +38,154 @@ public class MemberController {
-        if (foundMemberDto == null) {
-            response.setStatus(HttpStatus.NOT_FOUND.value());
+    @GetMapping("/my-info")
+    public ResponseEntity<MemberInfoDto> getMyInfo(@SessionAttribute(name = SessionConstants.loginMember, required = false) Long memberId) {
```
이런 부분들에 대해서 이해하고 있는지 하나하나 확인해보면, 좋을 것 같아요.
